### PR TITLE
Incorrect payment allocation bug

### PR DESCRIPTION
**Pull request by KietTruongSFDO**
# Critical Changes

# Changes

# Issues Closed

# New Metadata

# Deleted Metadata

### CODE REVIEW

**Comment by KietTruongSFDO** [comment]
@jbrandolinisf Any thoughts on this PR?

File: src/classes/ALLO_PaymentSync_TEST.cls [hunk 1]
```diff
@@ -618,6 +618,213 @@ private with sharing  class ALLO_PaymentSync_TEST {
             'Inserting or updating opportunity and payment allocations for the same opportunity failed to cause the expected exception.');
     }
 
+    /*******************************************************************************************************
```
[the remark this answers is not in the document]
Comment (KietTruongSFDO) [reply src/classes/ALLO_PaymentSync_TEST.cls]: These test cases were to cover the case that spawned this WI initially. The case where the Payment has the allocation set and then updating the Opportunity Allocation amount caused the payment allocation amount to not get set properly. 

So I added these test cases to make sure that setting the opportunity allocation will reset the payment allocation properly.

File: src/classes/ALLO_PaymentSync_TEST.cls [hunk 2]
```diff
@@ -618,6 +618,213 @@ private with sharing  class ALLO_PaymentSync_TEST {
...
+                                General_Accounting_Unit__c,
+                                General_Accounting_Unit__r.Name,
+                                Amount__c,
+                                Percent__c
+                       from Allocation__c
+                       where Payment__c = :payments[0].Id
+                       AND General_Accounting_Unit__c =: GAUId
+                       order by Amount__c DESC];
+
+        System.assert(paymentAllocation.Percent__c == 50, 'The allocation percent is incorrect. It should be 50, but it is actually: ' + paymentAllocation);
```
[the remark this answers is not in the document]
Comment (KietTruongSFDO) [reply src/classes/ALLO_PaymentSync_TEST.cls]: I'm just used to writing it out in the way I did. No reason in particular, I can change it to your format if you'd like

File: src/classes/ALLO_PaymentSync_SVC.cls [hunk 3]
```diff
@@ -241,7 +242,7 @@ public class ALLO_PaymentSync_SVC {
             }
 
             if (opportunity.Amount != null && opportunity.Amount != 0) {
-                context.scheduleRatio = totalPaymentsPaidUnpaid / opportunity.Amount;
+                context.scheduleRatio = context.totalUnpaidPayments / opportunity.Amount;
             }
```
[the remark this answers is not in the document]
Comment (KietTruongSFDO) [reply src/classes/ALLO_PaymentSync_SVC.cls]: Added a short comment, let me know if this is sufficient.

File: src/classes/ALLO_PaymentSync_TEST.cls (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbrandolinisf) [reply src/classes/ALLO_PaymentSync_TEST.cls]: Your call.  it was more or less feedback for the future, as it is less work to write it out the other way, with effectively the same result.  Do what you like with this, meaning update or don't, and mark it resolved.

[the remark this answers is not in the document]
Comment (KietTruongSFDO) [reply src/classes/ALLO_PaymentSync_TEST.cls]: Resolved with no changes!

File: src/classes/ALLO_PaymentSync_TEST.cls [hunk 4]
```diff
@@ -618,6 +618,213 @@ private with sharing  class ALLO_PaymentSync_TEST {
...
+                       where Payment__c = :payments[0].Id
+                       AND General_Accounting_Unit__c =: GAUId
+                       order by Amount__c DESC];
+
+        System.assert(paymentAllocation.Percent__c == 2, 'The allocation percent is incorrect. It should be 2, but it is actually: ' + paymentAllocation);
+        Test.stopTest();
+    }
+
+    /*******************************************************************************************************
+    * @description Test when an opportunity is underscheduled, has payment level allocations changed,
```
Comment (jbrandolinisf) [inline src/classes/ALLO_PaymentSync_TEST.cls]: Does not match the method.

File: src/classes/ALLO_PaymentSync_TEST.cls [hunk 5]
```diff
@@ -618,6 +618,213 @@ private with sharing  class ALLO_PaymentSync_TEST {
...
+                       where Payment__c = :payments[0].Id
+                       AND General_Accounting_Unit__c =: GAUId
+                       order by Amount__c DESC];
+
+        System.assert(paymentAllocation.Percent__c == 50, 'The allocation percent is incorrect. It should be 50, but it is actually: ' + paymentAllocation);
+        Test.stopTest();
+    }
+
+    /*******************************************************************************************************
+    * @description Test when an opportunity is underscheduled, has payment level allocations changed,
```
Comment (jbrandolinisf) [inline src/classes/ALLO_PaymentSync_TEST.cls]: Can you be a little more specific on each test's description, meaning enough that we know the difference between each of the tests.  The three tests you added, have the same description (even the overscheduled one).

File: src/classes/ALLO_PaymentSync_TEST.cls [hunk 6]
```diff
@@ -618,6 +618,213 @@ private with sharing  class ALLO_PaymentSync_TEST {
             'Inserting or updating opportunity and payment allocations for the same opportunity failed to cause the expected exception.');
```
Comment (jbrandolinisf) [inline src/classes/ALLO_PaymentSync_TEST.cls]: This is not part of this change, but do you mind moving this to a label?

Comment (KietTruongSFDO) [reply src/classes/ALLO_PaymentSync_TEST.cls]: @jbrandolinisf Should this be moved into a variable instead of a label? I don't think this is necessary as a label since this is for test class use only.

Comment (jbrandolinisf) [reply src/classes/ALLO_PaymentSync_TEST.cls]: Oh, right, never mind.  Leave it as is.
